src/test/java/org/broadinstitute/consent/http/service/UserServiceTest.java

```diff
@@ -849,13 +856,257 @@ void testFindUsersInJsonArrayInvalidKey() {
+    LibraryCard lc = new LibraryCard();
+    lc.setCreateUserId(signingOfficial.getUserId());
+    testUser.setLibraryCards(List.of(lc));
+    Institution institutionFromEmail = new Institution();
+    institutionFromEmail.setId(1);
+    testUser.setInstitution(institutionFromEmail);
+
+    Institution soInstitution = new Institution();
+    institutionFromEmail.setId(2);
+
+    when(userDAO.findUserById(signingOfficial.getUserId())).thenReturn(signingOfficial);
+    when(institutionService.findInstitutionForEmail(signingOfficial.getEmail())).thenReturn(soInstitution);
+    assertTrue(service.handleLibraryCardForUser(testUser));
+    verify(libraryCardDAO).deleteAllLibraryCardsByUser(testUser.getUserId());
+  }
+
+  static class InstitutionAndLibraryCardVariationsProvider implements ArgumentsProvider {
+    @Override
+    public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
+      User testUser = generateUser();
+      Institution institution1 = new Institution();
+      institution1.setId(1);
+      Institution institution2 = new Institution();
+      institution2.setId(2);
+      List<LibraryCard> libraryCards1 = List.of(new LibraryCard());
+      LibraryCard libraryCard2 = new LibraryCard();
+      libraryCard2.setCreateUserId(1);
+      return Stream.of(
+          Arguments.of(institution1, testUser, libraryCards1, institution1, true),
+          Arguments.of(institution1, testUser, List.of(libraryCard2), institution1, true),
+          Arguments.of(institution1, testUser, libraryCards1, institution2, true),
+          Arguments.of(institution1, testUser, libraryCards1, null, true),
+          Arguments.of(institution1, testUser, null, institution1, false),
+          Arguments.of(institution1, testUser, null, institution2, true),
+          Arguments.of(institution1, testUser, null, null, true),
+          Arguments.of(null, testUser, libraryCards1, institution2, true),
+          Arguments.of(null, testUser, libraryCards1, null, true),
+          Arguments.of(null, testUser, null, institution2, true),
+          Arguments.of(null, testUser, null, null, false)
+      );
+    }
+  }
+
+  @ParameterizedTest
+  @ArgumentsSource(InstitutionAndLibraryCardVariationsProvider.class)
+  void testEnforceInstitutionAndLibraryCardVariations(Institution institutionFromMap, User testUser, List<LibraryCard> cards, Institution institutionFromDatabase, boolean expectsUserMod) {
+    testUser.setLibraryCards(cards);
+    User alteredUser = new User();
+    alteredUser.setEmail(testUser.getEmail());
+    when(institutionService.findInstitutionForEmail(testUser.getEmail())).thenReturn(institutionFromMap);
+    when(institutionService.findInstitutionById(testUser.getInstitutionId())).thenReturn(institutionFromDatabase);
+    if (expectsUserMod) {
+      when(userDAO.findUserByEmail(testUser.getEmail())).thenReturn(alteredUser);
+      validateAlteredUserIsReturned(testUser, service.enforceInstitutionAndLibraryCardRules(testUser));
```

Comment:
Since the last line in `enforceInstitutionAndLibraryCardRules()` is `return findUserByEmail(user.getEmail())`, if the user is modified, this should instead use `assertEquals()`.
```suggestion
      assertEquals(testUser, service.enforceInstitutionAndLibraryCardRules(testUser));
```

Similarly the `else` case is `assertEquals(testUser, ...)`

Reply:
The User equals comparison only considers the userId property.  It would always be true.

The modification to the user we're expecting are properties that are outside of that equivalence rule.